fix(dashboard): allow clearing optional collection, type, and country fields

bqst opened the pull request:
## Summary

**What** — What changes are introduced in this PR?

This PR adds the ability to clear/unset optional fields in product and inventory forms. Specifically:
- Product collection and type selectors can now be cleared
- Country of origin fields can now be cleared in product, product variant, and inventory forms
- The `CountrySelect` component has been refactored to use `Combobox` internally, enabling search and clear functionality

**Why** — Why are these changes relevant or necessary?  

Currently, when a user assigns a collection, type, or country of origin to a product/variant/inventory item, there is no way to unset that value through the UI. Once selected, the field is locked to that value, forcing users to either:
- Keep the unwanted value
- Delete and recreate the resource
- Manually update via API

This is an UX issue for optional fields that should allow users to remove their selection.

**How** — How have these changes been implemented?

1. **CountrySelect Component Refactor**:
   - Replaced `Select` component with `Combobox` internally
   - Added `allowClear` prop support
   - Implemented proper value handling to convert between empty string (`""`) and `undefined`
   - Now provides search functionality as a bonus

2. **Product Organization Forms** (create & edit):
   - Added `allowClear` prop to `collection_id` and `type_id` Combobox fields
   - Added value normalization: `value={field.value || undefined}` 
   - Added onChange handler: `onChange={(value) => field.onChange(value || "")}`
   - Ensures cleared values convert to empty string to match default values

3. **Country of Origin Fields**:
   - Added `allowClear` prop to CountrySelect in:
     - Product variant edit form
     - Product attributes form (create & edit)
     - Inventory item form (create & edit)

**Testing** — How have these changes been tested, or how can the reviewer test the feature?

**Manual Testing Steps:**

1. **Product Collection/Type**:
   - Navigate to any product → Edit → Organization
   - Select a collection and/or type
   - Save the form
   - Edit organization again → Click the "X" button on collection/type
   - Verify the value clears and the field shows as empty
   - Save and verify the collection/type is removed from the product

2. **Country of Origin**:
   - Navigate to product → Edit → Attributes  
   - Select a country of origin
   - Save the form
   - Edit attributes again → Click the "X" button on country
   - Verify the country clears
   - Save and verify the country is removed

3. **Inventory Items**:
   - Create or edit an inventory item
   - Select a country of origin in the attributes section
   - Click the clear button
   - Verify the country field clears properly

**Expected Behavior:**
- Clear button (X icon) appears when a value is selected
- Clicking clear removes the selected value
- Field returns to empty state
- Form can be saved with the cleared value
- Value is properly removed from the resource

---

## Examples

### Using allowClear with Combobox

// Product collection selector - optional field
<Combobox
  {...field}
  value={field.value || undefined}
  onChange={(value) => field.onChange(value || "")}
  options={collections.options}
  allowClear  // Enables the clear button
/>### Updated CountrySelect component

// Before: Used Select component (no search, no clear)
<CountrySelect {...field} />

// After: Uses Combobox internally (searchable + clearable)
<CountrySelect {...field} allowClear />  // For optional fields
<CountrySelect {...field} />             // For required fields (no clear button)### Value Handling Pattern

// The pattern used for optional single-select Combobox fields:
<Combobox
  {...field}
  value={field.value || undefined}  // Convert "" to undefined for Combobox
  onChange={(value) => field.onChange(value || "")}  // Convert undefined to "" for form
  allowClear
/>This ensures proper conversion between:
- Form uses: `""` (empty string) or `null` 
- Combobox uses: `undefined` (empty) or `string` value

---

## Checklist

Please ensure the following before requesting a review:

- [x] I have added a **changeset** for this PR
    - Every non-breaking change should be marked as a **patch**
    - To add a changeset, run `yarn changeset` and follow the prompts
- [x] The changes are covered by relevant **tests**
- [x] I have verified the code works as intended locally
- [x] I have linked the related issue(s) if applicable

---

## Additional Context

This fix addresses a recurring UX issue where users had no way to unset optional fields once a value was selected, improving the overall form editing experience.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Enable clearing product type/collection and country-of-origin across product, variant, and inventory forms by refactoring CountrySelect to Combobox and adjusting Combobox value handling.
> 
> - **UI Inputs**:
>   - **`Combobox`**: Treat `""` as no value for single-select; only show clear when value is non-empty; minor visibility logic tweaks.
>   - **`CountrySelect`**: Refactored to use `Combobox`; adds search and `allowClear`; normalizes `value` and `onChange` between `""` and `undefined`.
> - **Forms**:
>   - **Product Organization (create/edit)**: `Combobox` for `type_id` and `collection_id` now `allowClear` with `value={field.value || ""}` and `onChange={(v) => field.onChange(v || "")}`.
>   - **Product Attributes & Variant Edit**: `origin_country` uses `CountrySelect` with `allowClear`; label marked optional where applicable.
>   - **Inventory (create/edit)**: `origin_country` now clearable via `CountrySelect` with `allowClear`.
> - **Release**: Add patch changeset for `@medusajs/dashboard`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e58c751d80bab1724766ba82c59d99016c2fb436. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

bqst committed:
🐛 Allow clearing optional collection and type fields in product organization form

bqst committed:
🐛 Allow clearing optional collection, type, and country select fields

bqst committed:
🔧 Add changeset

fPolic commented on packages/admin/dashboard/src/components/inputs/country-select/country-select.tsx:
```diff
@@ -6,47 +6,45 @@ import {
+    ? (value ? value.toLowerCase() : "")
+    : ""
+
+  const handleChange = (newValue: string | undefined) => {
```
discussion: wondering if we should reset filed in this case or set it to `null`. If we treat `undefined` as `""` we would need to implement a check for the fields that are "string | null" in each `handleSubmit`.

bqst replied:
Good catch! After investigating, I found that using `undefined` causes the Combobox to switch to uncontrolled mode (since `isValueControlled = controlledValue !== undefined`), which creates bugs where cleared values still display.

**Solution implemented:**

1. **Always use `""` (empty string) instead of `undefined`** to keep components controlled:
   ```typescript
   // In forms
   value={field.value || ""}  // Not undefined
   onChange={(value) => field.onChange(value || "")}
   ```

2. **Updated Combobox to treat `""` as "no value"** for single-select:
   ```typescript
   // In Combobox component
   const hasValue = isArrayValue 
     ? selectedValues?.length > 0 
     : selectedValues?.length > 0 && selectedValues !== ""
   ```

3. **Updated CountrySelect** similarly to maintain controlled state with `""`.

**Benefits:**
- ✅ Components stay controlled (no switching to uncontrolled mode)
- ✅ Clear functionality works correctly in all scenarios
- ✅ Forms continue converting `""` → `null` in `handleSubmit` as they already do
- ✅ No additional checks needed in each form handler

This approach maintains the existing pattern where forms use `""` internally and convert to `null` for the API.

fPolic reviewed (commented):
looks good overall @bqst!
have one discussion comment

bqst committed:
🐛 Optional fields stay controlled when cleared

bqst committed:
Merge branch 'develop' into fix/combobox-allow-clear

NicolasGorga commented on .changeset/lovely-suns-spend.md:
```diff
@@ -0,0 +1,5 @@
+"@medusajs/dashboard": patch
+---
+
+Add ability to clear optional collection, type, and country fields in product and inventory forms
```
Could you change it to match the PR title: `fix(dashboard): allow clearing optional collection, type, and country fields`

bqst replied:
Done, changeset aligned with the PR title in 27a21d2383.

NicolasGorga commented on packages/admin/dashboard/src/components/inputs/combobox/combobox.tsx:
```diff
@@ -312,7 +315,7 @@ const ComboboxImpl = <T extends Value = string>(
           />
         </div>
-        {allowClear && controlledValue && (
+        {allowClear && controlledValue && controlledValue !== "" && (
```
@bqst an empty string is falsey, so no need for the additional condition, as it will only pass the condition when it has a non empty string

bqst replied:
You're right, I left that redundancy in. Cleaned it up in 27a21d2383 (also simplified the `hasValue` check that had the same pattern).

NicolasGorga reviewed (changes requested):
Also, check the comment from the PR reviewer regarding the product attribute form and the variant and inventory ones as well, since they are sending the `origin_country` value as is if an empty string, without converting to NULL. It's not that it is going to make anything fail, but best for consistency's sake. Other than that LGTM!

NicolasGorga committed:
Merge branch 'develop' into fix/combobox-allow-clear

bqst committed:
refactor(dashboard): address PR review feedback for clearable fields

- combobox: drop redundant !== "" checks; empty string is already falsy
  so truthiness alone is enough.
- product attributes form and inventory item edit form: convert empty
  origin_country to null on submit, matching the convention used by
  variant edit (transformNullableFormData) and the rest of the API.
- changeset message aligned with the PR title.

bqst committed:
Merge remote-tracking branch 'upstream/develop' into fix/combobox-allow-clear

# Conflicts:
#	packages/admin/dashboard/src/components/inputs/country-select/country-select.tsx

shahednasser reviewed (approved):
Thanks!
